Enrich 4.0.0 (#749)
snowplow/documentation: docs/pipeline-components-and-applications/enrichment-components/configuration-reference/index.md

| `telemetry.userProvidedId` | Optional. See [here](/docs/getting-started-on-community-edition/telemetry/index.md#how-can-i-help) for more information. |
| `featureFlags.acceptInvalid` | Optional. Default: `false`. Enrich *3.0.0* introduces the validation of the enriched events against atomic schema before emitting. If set to `false`, a bad row will be emitted instead of the enriched event if validation fails. If set to `true`, invalid enriched events will be emitted, as before. |
| `featureFlags.legacyEnrichmentOrder` | Optional. Default: `false`. In early versions of `enrich-kinesis` and `enrich-pubsub` (>= *3.1.5*), the Javascript enrichment incorrectly ran before the currency, weather, and IP Lookups enrichments. Set this flag to true to keep the erroneous behavior of those previous versions. |

> | `validation.atomicFieldsLimits`      | Optional. Default: (#TODO Link to `application.conf` with defaults in repo...?) Configuration for custom maximum atomic fields (strings) length. It's a map-like structure with keys being atomic field names and values being their max allowed length.                                                 |

Reviewer: ```suggestion
| `validation.atomicFieldsLimits`      | Optional. For the defaults, see [here](https://github.com/snowplow/enrich/blob/master/modules/common/src/main/resources/reference.conf). Configuration for custom maximum atomic fields (strings) length. It's a map-like structure with keys being atomic field names and values being their max allowed length.                                                 |
```

| `telemetry.userProvidedId` | Optional. See [here](/docs/getting-started-on-community-edition/telemetry/index.md#how-can-i-help) for more information. |
| `featureFlags.acceptInvalid` | Optional. Default: `false`. Enrich *3.0.0* introduces the validation of the enriched events against atomic schema before emitting. If set to `false`, a bad row will be emitted instead of the enriched event if validation fails. If set to `true`, invalid enriched events will be emitted, as before. |
| `featureFlags.legacyEnrichmentOrder` | Optional. Default: `false`. In early versions of `enrich-kinesis` and `enrich-pubsub` (>= *3.1.5*), the Javascript enrichment incorrectly ran before the currency, weather, and IP Lookups enrichments. Set this flag to true to keep the erroneous behavior of those previous versions. |

> | `validation.atomicFieldsLimits` | Optional. For the defaults, see [here](https://github.com/snowplow/enrich/blob/master/modules/common/src/main/resources/reference.conf). Configuration for custom maximum atomic fields (strings) length. It's a map-like structure with keys being atomic field names and values being their max allowed length. |

Reviewer: Could you add `_(since 4.0.0)_`?

| `output.bad.topicName` | Optional. Name of the Kafka topic to write to |
| `output.bad.bootstrapServers` | Optional. A list of host:port pairs to use for establishing the initial connection to the Kafka cluster |
| `output.bad.producerConf` | Optional. Kafka producer configuration. See [the docs](https://kafka.apache.org/documentation/#producerconfigs) for all properties |

> | `blobStorage.s3` | Optional. Set to `true` if S3 client should be initialized to download enrichments assets. |

Reviewer: Same comment as above

| `blobStorage.gcs` | Optional. Set to `true` if GCS client should be initialized to download enrichments assets. |
| `blobStorage.azureStorage` | Optional. Azure Blob Storage client configuration. ABS client won't be enabled if it isn't given. |
| `blobStorage.azureStorage.accounts` | Array of accounts to download from Azure Blob Storage. |

> Example values for the storage accounts :

Reviewer: ```suggestion
Example values for the Azure storage accounts :
```

When we'll know that all our customers don't have any invalid enriched events any more, we'll remove the feature flags and it will be impossible to emit invalid enriched events.

> Since `4.0.0`, it is possible to configure the lengths of the atomic fields with `validation.atomicFieldsLimits` (see above).

Reviewer: Worth copy-pasting the example here I think — the one from the “upgrading” page.
